### ISSUE TITLE
Highlighting doesn't apply on previous content

### ISSUE THREAD

**Issue by aulneau** [outline/rich-markdown-editor#408]
Hi again :~)

So i've extended the refractor/prism highlighter to include a language definition we have for the `clarity` smart contracting language, and I noticed that highlighting works fine (for any language) when you add a code block and add some code, however once you leave the document and come back to it, it seems as though prism isn't running to highlight the code. It happens with all languages, not just the custom one I've added.

I think this has something to do with the fact that I'm using yjs and the nodes/marks are affected a bit differently once it's saved/reloaded via yjs.

@tommoor I'm curious if you've ran into this with your work to make outline collaborative? I've started playing around with trying to force `prism` to highlight on mount of the editor component, but it doesn't seem to be working. Before I refactor the highlighter plugin in the editor itself, I wanted to see if there might be some guidance you'd have.

| first load | second load |
|---------|--------------|
| <img width="871" alt="Screen Shot 2021-03-14 at 12 11 09 PM" src="https://user-images.githubusercontent.com/11803153/111077528-2e93f300-84bf-11eb-9e82-c356232afa62.png"> | <img width="861" alt="Screen Shot 2021-03-14 at 12 11 17 PM" src="https://user-images.githubusercontent.com/11803153/111077529-2f2c8980-84bf-11eb-819a-2acb8c7a4e29.png"> |

**Comment by tommoor** [comment]
Yea, there's a bug there – it's related to this "deferred" line:
https://github.com/outline/rich-markdown-editor/blob/main/src/plugins/Prism.ts#L106

The _idea_ is good, in that the initial editor rendering shouldn't be blocked by all of the code highlighting (which is relatively slow), it's a good idea to move it to a second render frame so the initial paint time is faster. But this implementation doesn't work too well – would happily accept a fix.

**Comment by aulneau** [comment]
Yep that makes sense.

I spent some time working on the prism stuff, but ultimately ended up making a [`Shiki`](https://github.com/shikijs/shiki) extension as I really love that package. I'd be interested to know if you'd be open to switching!

The shiki extension takes a highlighter param that is loaded by the app/client. To get the highlighter, a few assets are needed (wasm, languages) and I was not really sure about how to make an async prosemirror extension, and so I am using [`use-asset`](https://github.com/pmndrs/use-asset) to make the loading of the highlighter suspense-full and as such, it's always available to the editor because it loads before the editor :~) so far i am happy with the results! 

I think people can take whatever method they want to load the highlighting assets, this is just what works for me (since other parts of my app are using suspense already).

Lmk if you would be interested in a PR to make this an option in the editor, I can bring over the specific code from my fork to the main repo.

Here is a gif with the highlighting and a refresh! 

![shiki-rich-markdown-editor](https://user-images.githubusercontent.com/11803153/111083366-3c577180-84db-11eb-9057-1331bb5bc127.gif)

**Comment by tommoor** [comment]
> Lmk if you would be interested in a PR to make this an option in the editor, I can bring over the specific code from my fork to the main repo.

I don't think we want an option of two code formatters, I'll stick with Prism for now unless this approach offers some great benefits. PR incoming for the delayed rendering fix

**Comment by aulneau** [comment]
> I'll stick with Prism for now unless this approach offers some great benefits

Yeah that makes total sense! It works for my needs because I'm using `shiki` to render the code on the next/mdx reader app that goes along with this editor app i'm working on. It's mostly just personal preference, too; I prefer the output from the onigasm/vscode engine behind `shiki`. 

Your PR looks good from a quick read through, I'll give it a shot tomorrow and report back if it fixes the issue!
